ReactiveCocoa/NSObject+Intercepting.swift

```diff
@@ -6,30 +6,419 @@ extension Reactive where Base: NSObject {
+				//
+				// Example: KVO setters if the instance is swizzled by KVO before RAC
+				//          does.
+				if !class_respondsToSelector(subclass, interopAlias) {
+					let immediateMethod = class_getImmediateMethod(subclass, selector)
+
+					let immediateImpl: IMP? = immediateMethod.flatMap {
+						let immediateImpl = method_getImplementation($0)
+						return immediateImpl.flatMap { $0 != _rac_objc_msgForward ? $0 : nil }
+					}
+
+					if let impl = immediateImpl {
+						class_addMethod(subclass, interopAlias, impl, typeEncoding)
+					}
+				}
 			}
 
-			let (signal, observer) = Signal<(), NoError>.pipe()
-			let isSuccessful = base._rac_setupInvocationObservation(for: selector,
-			                                                        protocol: nil,
-			                                                        receiver: observer.send(value:))
-			precondition(isSuccessful)
+			let state = InterceptingState(lifetime: reactive.lifetime)
+			setAssociatedValue(state, forKey: alias.utf8Start)
+
+			// Start forwarding the messages of the selector.
+			_ = class_replaceMethod(subclass, selector, _rac_objc_msgForward, typeEncoding)
+			
+			return state.signal
+		}
+	}
+}
+
+/// Swizzle `realClass` to enable message forwarding for method interception.
+///
+/// - parameters:
+///   - realClass: The runtime subclass to be swizzled.
+private func enableMessageForwarding(_ realClass: AnyClass, _ selectorCache: SelectorCache) {
+	let perceivedClass: AnyClass = class_getSuperclass(realClass)
+
+	typealias ForwardInvocationImpl = @convention(block) (Unmanaged<NSObject>, AnyObject) -> Void
+	let newForwardInvocation: ForwardInvocationImpl = { objectRef, invocation in
+		let selector = invocation.selector!
+		let alias = selectorCache[main: selector]
+		let interopAlias = selectorCache[interop: selector]
+
+		defer {
+			if let state = objectRef.takeUnretainedValue().associatedValue(forKey: alias.utf8Start) as! InterceptingState? {
+				state.observer.send(value: invocation)
+			}
+		}
+
+		let method = class_getInstanceMethod(perceivedClass, selector)!
+		let typeEncoding = method_getTypeEncoding(method)
+
+		if class_respondsToSelector(realClass, interopAlias) {
+			// RAC has preserved an immediate implementation found in the runtime
+			// subclass that was supplied by an external party.
+			//
+			// As the KVO setter relies on the selector to work, it has to be invoked
+			// by swapping in the preserved implementation and restore to the message
+			// forwarder afterwards.
+			//
+			// However, the IMP cache would be thrashed due to the swapping.
 
-			lifetime.ended.observeCompleted(observer.sendCompleted)
-			map.setObject(signal, forKey: selectorName)
+			let interopImpl = class_getMethodImplementation(realClass, interopAlias)
+			let previousImpl = class_replaceMethod(realClass, selector, interopImpl, typeEncoding)
+			invocation.invoke()
+			_ = class_replaceMethod(realClass, selector, previousImpl, typeEncoding)
```

Comment:
I wonder if it'd make sense to wrap some of the objc runtime APIs in a more Swift-friendly package internally to make the actual logic here more readable. 🤔